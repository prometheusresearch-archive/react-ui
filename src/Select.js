/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import React from 'react';

export const NOVALUE = '__NOVALUE__';

const NOVALUE_OPTION = {label: '', value: NOVALUE};

export default class Select extends React.Component {

  render() {
    let {options, value, allowNoValue} = this.props;
    if (value == null) {
      value = NOVALUE;
    }
    if (value === NOVALUE || allowNoValue) {
      options = [NOVALUE_OPTION].concat(options);
    }
    return (
      <select value={value} onChange={this.onChange}>
        {options.map(this.renderOption, this)}
      </select>
    );
  }

  renderOption(option) {
    let key = option.value === NOVALUE ? '__NOVALUE__' : option.value;
    return (
      <option key={key} value={option.value}>
        {option.label}
      </option>
    );
  }

  onChange = event => {
    let value = event.target.value;
    if (value === NOVALUE) {
      value = null;
    }
    this.props.onChange(value);
  }
}


