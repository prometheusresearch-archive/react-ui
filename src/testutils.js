/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import React from 'react';
import {isEqual} from 'lodash/lang';
import indent from 'indent-string';

import {
  findWithType as _findWithType,
  findAllWithType as _findAllWithType,
  getMountedInstance,
  findAll
} from 'react-shallow-testutils';

import {
  createRenderer as _createRenderer
} from 'react-addons-test-utils';

import _inspectElement, {
  inspectReactProp as _inspectReactProp
} from 'inspect-react-element';


import assert from 'power-assert';
import {stub, spy, mock} from 'sinon';

export {stub, spy, mock, assert};

class FormDataMock {

  append = spy();
}

let originalFormData = global.FormData;

export function mockFormData() {
  global.FormData = FormDataMock;
}

export function unmockFormData() {
  global.FormData = originalFormData;
}

function inspectReactProps(node) {
  return '__originalConfig' in node.props ?
    {...node.props.__originalConfig, children: node.props.children} :
    node.props || {};
}

function inspectReactProp(propName, propValue) {
  if (propName === '__originalConfig' || propName === 'key' || propName === 'ref') {
    return null;
  }
  return _inspectReactProp(propName, propValue);
}

export function inspectElement(node) {
  return _inspectElement(node, {inspectReactProps, inspectReactProp});
}

class ReactShallowRenderer {

  constructor() {
    this._renderer = _createRenderer();
  }

  render(element, context) {
    return this._renderer.render(element, context);
  }

  get instance() {
    return getMountedInstance(this._renderer);
  }

  get element() {
    return this._renderer.getRenderOutput();
  }

  findWithType(type) {
    return findWithType(this.element, type);
  }

  findAllWithType(type) {
    return findAllWithType(this.element, type);
  }

  assertElementWithType(type) {
    this.findWithType(type);
  }

  assertNoElementWithType(type) {
    assert(this.findAllWithType(type).length === 0);
  }


  findWithTypeProps(type, props) {
    return findWithTypeProps(this.element, type, props);
  }

  findAllWithTypeProps(type, props) {
    return findAllWithTypeProps(this.element, type, props);
  }

  assertElementWithTypeProps(type, props) {
    this.findWithTypeProps(type, props);
  }

  assertNoElementWithTypeProps(type, props) {
    assert(this.findAllWithTypeProps(type, props).length === 0);
  }


  findWithElement(element) {
    return findWithElement(this.element, element);
  }

  findAllWithElement(element) {
    return findAllWithElement(this.element, element);
  }

  assertElement(element) {
    this.findWithElement(element);
  }

  assertNoElement(element) {
    assert(this.findAllWithElement(element).length === 0);
  }

}

export function createRenderer() {
  return new ReactShallowRenderer();
}

export function findWithType(root, type) {
  let elements = findAll(root, _makeMatchWithType(type));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one partial match for element:
${indent(inspectElement(React.createElement(type)), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithType(root, type) {
  return findAll(root, _makeMatchWithType(type));
}

function _makeMatchWithType(type) {
  return function _matchWithType(element) {
    return element && element.type && (
      element.type === type ||
      element.type.Component === type
    );
  };
}

export function findWithTypeProps(root, type, props) {
  let elements = findAll(root, _makeMatchWithElement(type, props));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one (${elements.length}) partial match for element:
${indent(inspectElement(React.createElement(type, props)), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithTypeProps(root, type, props) {
  return findAll(root, _makeMatchWithElement(type, props));
}

export function findWithElement(root, spec) {
  spec = _normalizeReactElementSpec(spec);
  let elements = findAll(root, _makeMatchWithElement(spec.type, spec.props));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one (${elements.length}) partial match for element:
${indent(inspectElement(spec), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithElement(root, spec) {
  spec = _normalizeReactElementSpec(spec);
  return findAll(root, _makeMatchWithElement(spec.type, spec.props));
}

function _makeMatchWithElement(type, props) {
  let matchWithType = _makeMatchWithType(type);

  let matchWithProps = function(element) {
    for (let key in props) {
      if (!props.hasOwnProperty(key)) {
        continue;
      }
      if (!element.props.hasOwnProperty(key)) {
        continue;
      }
      if (key === '__originalConfig') {
        continue;
      }
      if (!isEqual(element.props[key], props[key])) {
        return false;
      }
    }
    return true;
  };
  return function _matchWithElement(element) {
    return element && matchWithType(element) && matchWithProps(element);
  };
}

function _normalizeReactElementSpec(spec) {
  let {props: {__originalConfig = null}} = spec;
  if (__originalConfig) {
    return {...spec, props: __originalConfig};
  } else {
    return spec;
  }
}

let _createElement = React.createElement;

function createElement(type, config, ...children) {
  config = {...config, __originalConfig: config || {}};
  let element = _createElement(type, config, ...children);
  return element;
}

React.createElement = createElement;

