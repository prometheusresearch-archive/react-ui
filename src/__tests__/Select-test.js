/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import React from 'react';

import {createRenderer, spy, assert} from '../testutils';
import Select, {NOVALUE} from '../Select';

describe('react-ui', function() {

  let options = [
    {value: 'a', label: 'A'},
    {value: 'b', label: 'B'}
  ];

  let renderer = null;

  beforeEach(function() {
    renderer = createRenderer();
  });

  describe('<Select />', function() {

    it('renders', function() {
      let onChange = spy();
      let elements;
      renderer.render(
        <Select
          onChange={onChange}
          options={options}
          />
      );

      elements = renderer.findAllWithType('option');
      assert(elements.length === 3);
      assert(elements[0].props.value === NOVALUE);
      assert(elements[1].props.value === 'a');
      assert(elements[2].props.value === 'b');

      assert(renderer.element.props.value === NOVALUE);

      assert(renderer.element.props.onChange);
      renderer.element.props.onChange({target: {value: 'a'}});

      assert(onChange.callCount === 1);
      assert(onChange.lastCall.args[0] === 'a');

      renderer.render(
        <Select
          value="a"
          onChange={onChange}
          options={options}
          />
      );

      elements = renderer.findAllWithType('option');
      assert(elements.length === 2);
      assert(elements[0].props.value === 'a');
      assert(elements[1].props.value === 'b');

      assert(renderer.element.props.value === 'a');

      renderer.render(
        <Select
          allowNoValue
          value="a"
          onChange={onChange}
          options={options}
          />
      );

      elements = renderer.findAllWithType('option');
      assert(elements.length === 3);
      assert(elements[0].props.value === NOVALUE);
      assert(elements[1].props.value === 'a');
      assert(elements[2].props.value === 'b');

      assert(renderer.element.props.value === 'a');

      renderer.element.props.onChange({target: {value: NOVALUE}});
      assert(onChange.callCount === 2);
      assert(onChange.lastCall.args[0] === null);
    });
  });

});


