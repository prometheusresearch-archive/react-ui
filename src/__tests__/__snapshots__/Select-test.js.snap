// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<select
  onChange={[Function]}
  value="__NOVALUE__"
>
  <option
    value="__NOVALUE__"
  >
    
  </option>
  <option
    value="a"
  >
    A
  </option>
  <option
    value="b"
  >
    B
  </option>
</select>
`;

exports[`renders 2`] = `
<select
  onChange={[Function]}
  value="a"
>
  <option
    value="a"
  >
    A
  </option>
  <option
    value="b"
  >
    B
  </option>
</select>
`;

exports[`renders 3`] = `
<select
  onChange={[Function]}
  value="a"
>
  <option
    value="__NOVALUE__"
  >
    
  </option>
  <option
    value="a"
  >
    A
  </option>
  <option
    value="b"
  >
    B
  </option>
</select>
`;
